File: ios/Spandan/Sources/Signal/SignalBuffer.swift
import Foundation

/// Rolling time-windowed buffer of ROI-averaged RGB samples. Real, permanent
/// code -- bookkeeping only (drop samples older than the window), independent
/// of whichever algorithm consumes the buffer. Direct port of
/// android/app/.../signal/SignalBuffer.kt.
final class SignalBuffer {

    /// The single, tunable knob for how much wall-clock history the rolling
    /// buffer keeps. Kept at the same value the Android port settled on after
    /// its own on-device timing/window-length investigation (see
    /// android/README.md's "Window-length change: 10s -> 25s" section) --
    /// this iOS port has not repeated that investigation on real iPhone
    /// hardware (different camera/ISP throughput than the Android port's
    /// Galaxy A35), so 25s here is inherited, not independently re-verified.
    static let windowDurationSeconds: Double = 25.0

    private let windowSeconds: Double
    private var samples: [RgbSample] = []
    private let lock = NSLock()

    init(windowSeconds: Double = SignalBuffer.windowDurationSeconds) {
        self.windowSeconds = windowSeconds
    }

    func add(_ sample: RgbSample) {
        lock.lock()
        defer { lock.unlock() }
        samples.append(sample)
        let cutoffMs = sample.timestampMs - Int64(windowSeconds * 1000)
        while let first = samples.first, first.timestampMs < cutoffMs {
            samples.removeFirst()
        }
    }

    func snapshot() -> [RgbSample] {
        lock.lock()
        defer { lock.unlock() }
        return samples
    }

    func clear() {
        lock.lock()
        defer { lock.unlock() }
        samples.removeAll()
    }
}

File: android_segment30_mediapipe/app/src/main/java/com/spandan/app/signal/SignalBuffer.kt
package com.spandan.app.signal

/**
 * Rolling time-windowed buffer of ROI-averaged RGB samples. Real, permanent
 * code -- this is bookkeeping (drop samples older than the window),
 * independent of whichever algorithm eventually consumes the buffer.
 */
class SignalBuffer(private val windowSeconds: Double = WINDOW_DURATION_SECONDS) {

    private val samples = ArrayDeque<RgbSample>()

    @Synchronized
    fun add(sample: RgbSample) {
        samples.addLast(sample)
        val cutoffMs = sample.timestampMs - (windowSeconds * 1000).toLong()
        while (samples.isNotEmpty() && samples.first().timestampMs < cutoffMs) {
            samples.removeFirst()
        }
    }

    @Synchronized
    fun snapshot(): List<RgbSample> = samples.toList()

    @Synchronized
    fun clear() = samples.clear()

    companion object {
        /**
         * The single, tunable knob for how much wall-clock history the rolling
         * buffer keeps. Everything downstream (FFT bin resolution, on-screen
         * jitter, responsiveness to a changing HR) is a function of this one
         * number -- see android/README.md's "Diagnostic" and window-length
         * verification sections for the measurements behind the current value.
         *
         * Was 10.0 (~134 samples at the measured ~13.4Hz on-device throughput).
         * Raised to 25.0 (~335 samples) after the timing diagnostic ruled out
         * bursty/irregular sampling (CV=0.063, zero gaps >2x mean) as the cause
         * of on-screen jitter, leaving "too few samples per window" as the
         * better-supported explanation. 25s is a deliberate middle ground, not
         * an attempt to match MATLAB's ~80s/~2400-sample validated clips
         * (which would need ~3 minutes of steady holding at this throughput --
         * impractical for a live demo).
         */
        const val WINDOW_DURATION_SECONDS: Double = 25.0
    }
}
